buildSrc/subprojects/configuration/src/main/kotlin/org/gradle/gradlebuild/java/AvailableJavaInstallations.kt

```diff
@@ -66,71 +63,83 @@ private
-        javaInstallationForTest = when (javaHomeForTest) {
-            null -> currentJavaInstallation
-            else -> detectJavaInstallation(javaHomeForTest)
+        javaInstallationForTest = determineJavaInstallation(testJavaHomePropertyName)
+        javaInstallationForCompilation = determineJavaInstallationForCompilation()
+    }
+
+    private
+    fun determineJavaInstallationForCompilation() = if (JavaVersion.current().isJava9Compatible) currentJavaInstallation else determineJavaInstallation(java9HomePropertyName)
+
+    private
+    fun determineJavaInstallation(propertyName: String): JavaInstallation {
+        val resolvedJavaHome = resolveJavaHomePath(propertyName)
+        when (resolvedJavaHome) {
+            null -> return currentJavaInstallation
+            else -> return detectJavaInstallation(resolvedJavaHome)
         }
     }
 
-    fun jdkForCompilation(javaVersion: JavaVersion) = when {
-        javaVersion == currentJavaInstallation.javaVersion -> currentJavaInstallation
-        javaInstallations.containsKey(javaVersion) -> javaInstallations[javaVersion]!!
-        javaVersion <= currentJavaInstallation.javaVersion -> currentJavaInstallation
-        else -> throw IllegalArgumentException("No Java installation found which supports Java version $javaVersion")
+    fun validateForAllBuilds() {
+        validate(validateBuildJdks())
     }
 
-    fun validateForRemoteBuildCacheUsage() {
-        val validationErrors = validateCompilationJdks()
-        if (validationErrors.isNotEmpty()) {
-            val message = formatValidationError(
-                "In order to have cache hits from the remote build cache, your environment needs to be configured accordingly!",
-                validationErrors
-            )
-            throw GradleException(message)
+    fun validateForCompilation() {
+        if (remoteBuildCacheEnabled()) {
+            validate(validateForRemoteCache())
         }
+        validate(validateCompilationJdks())
     }
 
     fun validateForProductionEnvironment() {
-        val validationErrors = validateCompilationJdks() +
-            mapOf(
-                validationMessage(testJavaHomePropertyName, javaInstallationForTest, oracleJdk8) to (javaInstallationForTest.vendorAndMajorVersion != oracleJdk8)
-            ).filterValues { it }.keys
-        if (validationErrors.isNotEmpty()) {
-            throw GradleException(formatValidationError("JDKs not configured correctly for production build.", validationErrors))
+        validate(validateProductionJdks())
+    }
+
+    private
+    fun remoteBuildCacheEnabled() = (project.gradle as GradleInternal).settings.buildCache.remote?.isEnabled == true
+
+    private
+    fun validateForRemoteCache(): Map<String, Boolean> =
+        mapOf("Must use Oracle JDK 8/9 to perform this build. Is currently ${currentJavaInstallation.vendorAndMajorVersion} at ${currentJavaInstallation.javaHome}." to
```

Comment:
We should print something here which relates the error message to the remote cache usage. So the user knows it is possible to switch off the remote cache to run the build with the given JDK.

Reply:
Done.